src/drawing/polygon.rs

```diff
@@ -99,33 +97,26 @@ where
         }
 
         intersections.sort();
-        let mut i = 0;
-        loop {
-            // Handle points where multiple lines intersect
-            while i + 1 < intersections.len() && intersections[i] == intersections[i + 1] {
-                i += 1;
-            }
-            if i >= intersections.len() {
-                break;
-            }
-            if i + 1 == intersections.len() {
-                draw_if_in_bounds(canvas, intersections[i], y, color);
-                break;
-            }
-            let from = max(0, min(intersections[i], width as i32 - 1));
-            let to = max(0, min(intersections[i + 1], width as i32 - 1));
-            for x in from..to + 1 {
-                canvas.draw_pixel(x as u32, y as u32, color);
+        intersections.chunks(2).for_each(|range| {
+            let mut from = min(range[0], width as i32);
+            let mut to = min(range[1], width as i32 - 1);
+            if from < width as i32 && to >= 0 {
+                // draw only if range appears on the canvas
+                from = max(0, from);
+                to = max(0, to);
+
+                for x in from..to + 1 {
+                    canvas.draw_pixel(x as u32, y as u32, color);
+                }
             }
-            i += 2;
-        }
+        });
 
         intersections.clear();
     }
 
-    for edge in &edges {
+    edges.iter().for_each(|edge| {
```

Comment:
Nit: this is also a non-functional tweak that doesn't help with code readability. In this case the old code is also more idiomatic.